Apromore-Extras/Build-Tools/src/main/resources/checkstyle.xml

```diff
@@ -1,256 +1,223 @@
+  http://checkstyle.sf.net (or in your downloaded distribution).
+
+  Most Checks are configurable, be sure to consult the documentation.
+
+  To completely disable a check, just comment it out or delete it from the file.
+
+  Finally, it is worth reading the documentation.
+
+-->
+
+<!--
+
+  Comparison to the draft Apromore coding style at
+  https://apromore.atlassian.net/wiki/spaces/AP/pages/450691073/Apromore+coding+standards
+
+  1. Code formatting
+
+    a. The length of code lines should be within one page width
+       -> LineLength
+
+    b. Use indentation for code layout (replace tabs with 4 4 spaces)
+       -> FileTabCharacter
+
+  3. Functions/methods
+
+    a. Function should be small and cohesive. A one-page long function is not small.
+       -> MethodLength
+
+    d. Minimize the number of function arguments
+       -> ParameterNumber
+
+  5. Comments
+
+    c. Each file should have at least a license header and comment explaining the purpose, usage and/or relationship with other classes (look at the JDK)
+       -> Header (but we'll be using maven-license-plugin instead)
+       -> MissingJavadocType (see https://checkstyle.sourceforge.io/config_javadoc.html#MissingJavadocType)
+
+-->
+
+<module name="Checker">
+    <!--
+        If you set the basedir property below, then all reported file
+        names will be relative to the specified directory. See
+        https://checkstyle.org/5.x/config.html#Checker
+
+        <property name="basedir" value="${basedir}"/>
+    -->
+
+    <property name="fileExtensions" value="java, properties, xml"/>
+
+    <!-- Checks that a package-info.java file exists for each package.     -->
+    <!-- See http://checkstyle.sf.net/config_javadoc.html#JavadocPackage -->
+    <!-- <module name="JavadocPackage"/> -->
+
+    <!-- Checks whether files end with a new line.                        -->
+    <!-- See http://checkstyle.sf.net/config_misc.html#NewlineAtEndOfFile -->
+    <module name="NewlineAtEndOfFile"/>
+
+    <!-- Checks that property files contain the same keys.         -->
+    <!-- See http://checkstyle.sf.net/config_misc.html#Translation -->
+    <module name="Translation"/>
+
+    <!-- Checks for Size Violations.                    -->
+    <!-- See http://checkstyle.sf.net/config_sizes.html -->
+    <module name="LineLength">
+        <property name="max" value="120"/>
+    </module>
+    <module name="FileLength"/>
+
+    <!-- Checks for whitespace                               -->
+    <!-- See http://checkstyle.sf.net/config_whitespace.html -->
+    <module name="FileTabCharacter"/>
+
+    <!-- Miscellaneous other checks.                   -->
+    <!-- See http://checkstyle.sf.net/config_misc.html -->
+    <module name="RegexpSingleline">
+       <property name="format" value="\s+$"/>
+       <property name="minimum" value="0"/>
+       <property name="maximum" value="0"/>
+       <property name="message" value="Line has trailing spaces."/>
+    </module>
```

Comment:
This rule prevents adding an empty line to format code blocks for logical flow and readability. The message "Line has trailing spaces" is misleading in this case. I think it aims to check non-empty lines with trailing spaces, not empty lines?

Reply:
You can add an empty line, but that line can't have whitespace to match the surrounding indentation; it has to be genuinely empty (i.e. just a newline).  Between this and disallowing tabs, you ensure that there's never any invisible whitespace.

Reply:
On the other hand, consecutive empty lines should be checked but it seems there are no rules.

Reply:
"You can add an empty line, but that line can't have whitespace to match the surrounding indentation; it has to be genuinely empty (i.e. just a newline)": Tested and it doesn't pass that rule with the newline only. See screenshot.

![Uploading image.png…]()

Reply:
Revised: my Eclipse automatically adds whitespaces for indentation. Then, it would need a little hassle to pass this rule.